File: 3569-count-of-substrings-containing-every-vowel-and-k-consonants-ii/count-of-substrings-containing-every-vowel-and-k-consonants-ii.cpp
class Solution {
public:
    bool check(char c)
    {
        return c == 'a' || c == 'e' || c == 'i' || c == 'o'|| c == 'u';
    }

    long long greaterThanEqualToKConsonants(string word, int k)
    {
        int n = word.size();
        unordered_map<char, int> vowelMap;
        long long count = 0, consonants = 0, i = 0;
        for(int j = 0; j < n; j++)
        {
            char ch = word[j];

            if(check(ch))
            vowelMap[ch]++;

            else
            consonants++;

            while(i <= j && vowelMap.size() == 5 && consonants >= k)
            {
                count += (n - j);

                char ithChar = word[i];
                if(check(ithChar))
                {
                    vowelMap[ithChar]--;
                    if(vowelMap[ithChar] == 0)
                    vowelMap.erase(ithChar);
                }
                else
                {
                    consonants--;
                }

                i++;
            }
        }
        return count;
    }

    long long countOfSubstrings(string word, int k) 
    {
        return greaterThanEqualToKConsonants(word, k) - greaterThanEqualToKConsonants(word, k + 1);
    }
};

File: 3569-count-of-substrings-containing-every-vowel-and-k-consonants-ii/count-of-substrings-containing-every-vowel-and-k-consonants-ii.java
class Solution {
    public boolean check(char c) 
    {
        return c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u';
    }

    public long greaterThanEqualToKConsonants(String word, int k) 
    {
        int n = word.length();
        Map<Character, Integer> vowelMap = new HashMap<>();
        long count = 0;
        int consonants = 0;
        int i = 0;

        for (int j = 0; j < n; j++) 
        {
            char ch = word.charAt(j);

            if(check(ch)) 
            vowelMap.put(ch, vowelMap.getOrDefault(ch, 0) + 1);
            else 
            consonants++;

            while (i <= j && vowelMap.size() == 5 && consonants >= k) 
            {
                count += (n - j);

                char ithChar = word.charAt(i);
                if (check(ithChar)) 
                {
                    vowelMap.put(ithChar, vowelMap.get(ithChar) - 1);
                    if(vowelMap.get(ithChar) == 0)
                    vowelMap.remove(ithChar);
                }

                else 
                consonants--;

                i++;
            }
        }
        return count;
    }

    public long countOfSubstrings(String word, int k) 
    {
        return greaterThanEqualToKConsonants(word, k) - greaterThanEqualToKConsonants(word, k + 1);
    }
}
